flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/feature/TfIdfTransformer.scala

```diff
@@ -0,0 +1,128 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.feature
+
+import org.apache.flink.api.scala.{DataSet, _}
+import org.apache.flink.ml.common.{Parameter, ParameterMap, Transformer}
+import org.apache.flink.ml.math.SparseVector
+
+import scala.collection.mutable.LinkedHashSet
+import scala.math.log;
+
+/**
+ * This transformer calcuates the term-frequency times inverse document frequency for the give DataSet of documents.
+ * The DataSet will be treated as the corpus of documents. The single words will be filtered against the regex:
+ * <code>
+ * (?u)\b\w\w+\b
+ * </code>
+ * <p>
+ * The TF is the frequence of a word inside one document
+ * <p>
+ * The IDF for a word is calculated: log(total number of documents / documents that contain the word) + 1
+ * <p>
+ * This transformer returns a SparseVector where the index is the hash of the word and the value the tf-idf.
+ * @author Ronny Bräunlich
+ * @author Vassil Dimov
+ * @author Filip Perisic
+ */
+class TfIdfTransformer extends Transformer[(Int, Seq[String]), (Int, SparseVector)] {
+
+  override def transform(input: DataSet[(Int /* docId */ , Seq[String] /*The document */ )], transformParameters: ParameterMap): DataSet[(Int, SparseVector)] = {
```

Comment:
We had some discussion about the type in [FLINK-1999](https://issues.apache.org/jira/browse/FLINK-1999) and since there were no objections we sticked to Seq[String]. With Seq[String] the caller has to decide how to split his text.